src/runtime/synchronization_common.h

```diff
@@ -0,0 +1,934 @@
+        } else {
+          *data_location = data->next;
+          prev = data;
+          data = next;
+        }
+    }
+
+    bucket.mutex.unlock();
+
+    for (size_t i = 0; i < waiters; i++) {
+        temp_list[i]->parker.unpark();
+    }
+
+    // TODO: decide if this really needs to be two loops.
+    for (size_t i = 0; i < waiters; i++) {
+        temp_list[i]->parker.unpark_finish();
+    }
+
+    if (temp_list != &temp_list_storage[0]) {
+        free(temp_list);
+    }
+
+    return waiters;
+}
+
+int unpark_requeue(uintptr_t addr_from, uintptr_t addr_to, parking_control &control, uintptr_t unpark_info) {
+    bucket_pair buckets = lock_bucket_pair(addr_from, addr_to);
+
+    validate_action action;
+    if (!control.validate(&control, action)) {
+        unlock_bucket_pair(buckets);
+        return 0;
+    }
+
+    queue_data **data_location = &buckets.from.head;
+    queue_data *prev = NULL;
+    queue_data *data = *data_location;
+    queue_data *requeue = NULL;
+    queue_data *requeue_tail = NULL;
+    queue_data *wakeup = NULL;
+
+    while (data != NULL) {
+        uintptr_t cur_addr;
+        __atomic_load(&data->sleep_address, &cur_addr, __ATOMIC_RELAXED);
+
+        queue_data *next = data->next;
+        if (cur_addr == addr_from) {
+            *data_location = next;
+
+            if (buckets.from.tail == data) {
+                buckets.from.tail = prev;
+            }
+
+            if (action.unpark_one && wakeup == NULL) {
+                wakeup = data;
+            } else {
+                if (requeue == NULL) {
+                    requeue = data;
+                } else {
+                    requeue_tail->next = data;
+                }
+
+                requeue_tail = data;
+                __atomic_store(&data->sleep_address, &addr_to, __ATOMIC_RELAXED);
+            }
+            data = next;
+            // TODO: prev ptr?
+        } else {
+            data_location = &data->next;
+            prev = data;
+            data = next;
+        }
+    }
+
+    if (requeue != NULL) {
+        requeue_tail->next = NULL;
+        if (buckets.to.head == NULL) {
+            buckets.to.head = requeue;
+        } else {
+            buckets.to.tail->next = requeue;
+        }
+        buckets.to.tail = requeue_tail;
+    }
+
+    control.requeue_callback(&control, action, wakeup != NULL, requeue != NULL);
+
+    if (wakeup != NULL) {
+        wakeup->unpark_info = unpark_info;
+        wakeup->parker.unpark_start();
+        unlock_bucket_pair(buckets);
+        wakeup->parker.unpark();
+        wakeup->parker.unpark_finish();
+    } else {
+        unlock_bucket_pair(buckets);
+    }
+
+    return wakeup != NULL && action.unpark_one;
+}
+
+struct mutex_parking_control : parking_control {
+    uintptr_t *lock_state;
+
+    mutex_parking_control(uintptr_t *lock_state);
+};
+
+// Only used in parking -- lock_full.
+bool mutex_parking_control_validate(void *control, validate_action &action) {
```

Comment:
Should these be members of mutex_parking_control?

Reply:
I guess I could make them static members. Originally they were virtual members, but we can't use vtables in the runtime. The parking control dispatch is the only part of this I'm not happy with re: ease of understanding and code maintainability.